core/pkg/monitor/dcgm_exporter.go

```diff
@@ -0,0 +1,343 @@
+		params.Client.RetryWaitMin = DefaultOpenMetricsRetryWaitMin
+		params.Client.RetryWaitMax = DefaultOpenMetricsRetryWaitMax
+		params.Client.HTTPClient.Timeout = DefaultOpenMetricsTimeout
+		params.Client.Backoff = clients.ExponentialBackoffWithJitter
+	}
+
+	de := &DCGMExporter{logger: params.Logger}
+
+	// Case 1: /metrics endpoint.
+	if strings.HasSuffix(params.URL, "/metrics") {
+		// TODO: Implement
+		return nil
+	}
+
+	// Case 2: Prometheus API (/api/v1/query) query.
+	var err error
+
+	// Split the URL into the base URL and the query.
+	de.url, de.queries, err = ParsePromQLURL(params.URL)
+	if err != nil {
+		de.logger.Error("monitor: openmetrics: error parsing URL", "error", err)
+		return nil
+	}
+	de.logger.Debug("monitor: openmetrics: parsed URL", "url", de.url, "queries", de.queries)
+
+	// Create headers config.
+	headersConfig := &config.Headers{
+		Headers: make(map[string]config.Header),
+	}
+	for key, value := range params.Headers {
+		headersConfig.Headers[key] = config.Header{
+			Values: []string{value},
+		}
+	}
+
+	// Create a chain of RoundTrippers: headers -> retryable client.
+	roundTripper := config.NewHeadersRoundTripper(
+		headersConfig,
+		params.Client.StandardClient().Transport,
+	)
+
+	apiClient, err := api.NewClient(api.Config{
+		Address:      de.url,
+		RoundTripper: roundTripper,
+	})
+	if err != nil {
+		de.logger.Error("monitor: dcgm_exporter: error creating Prometheus API client", "error", err)
+		return nil
+	}
+
+	// Create the Prometheus API client.
+	de.v1api = v1.NewAPI(apiClient)
+
+	return de
+}
+
+func (de *DCGMExporter) Queries() []string {
+	return de.queries
+}
+
+func (de *DCGMExporter) Name() string {
+	return "dcgm_exporter"
+}
+
+// Sample fetches and parses the metrics from the endpoint and returns them as a StatsRecord.
+func (de *DCGMExporter) Sample() (*spb.StatsRecord, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultOpenMetricsTimeout)
+	defer cancel()
+
+	metrics := make(map[string]any)
+
+	for _, query := range de.queries {
+		result, warnings, err := de.v1api.Query(
+			ctx,
+			query,
+			time.Now(),
+			v1.WithTimeout(DefaultOpenMetricsTimeout),
+		)
+		if err != nil {
+			de.logger.Error("monitor: dcgm_exporter: error querying Prometheus API endpoint", "error", err)
+			return nil, err
+		}
+		if len(warnings) > 0 {
+			de.logger.Warn("monitor: openmetrics: warnings querying Prometheus API endpoint", "warnings", warnings)
+		}
+
+		// Process the results based on type
+		vector, ok := result.(model.Vector)
+		if !ok {
+			de.logger.Error("monitor: dcgm: unexpected result type", "type", fmt.Sprintf("%T", result))
+			continue
+		}
+
+		// Process each sample in the vector
+		for _, sample := range vector {
+			metricName := string(sample.Metric["__name__"])
+			labels := make(map[string]string)
+
+			for labelName, labelValue := range sample.Metric {
+				if labelName != "__name__" {
+					labels[string(labelName)] = string(labelValue)
+				}
+			}
```

Comment:
maybe helper function that returns the relevant metric name, gpu index and nodes?

Reply:
could be a struct object that is easy to access